### PR TITLE
Rework backend scheme

### PR DESCRIPTION
**Pull request by ghjm**
This PR makes the following changes:

- Instead of `RunBackend` calling `backend.Start()` which then calls back a `BackendSessionFunc`, backends register with a simplified `AddBackend` call, allowing more logic to be abstracted in the framework rather than requiring implementation in each backend.
- Utility functions are added for the common use cases of a Dial/redial and Listen/Accept backend types.
- Netceptor is restructured to use Go contexts instead of the shutdown channels slice, allowing for cleaner and more reliable shutdown, as well as being more idiomatic to Go.
- The logger is updated to treat log levels case insensitively, so users can run `--log-level debug` (lower case D) without error.

See https://github.com/project-receptor/receptor/issues/28

### CODE REVIEW

File: pkg/netceptor/netceptor.go
Comment (squidboylan) [inline pkg/netceptor/netceptor.go:637]: im not sure i understand why this isnt a `logger.Info` output

File: example/net.go
Comment (squidboylan) [inline example/net.go]: I think this can be `context.Background()` since the netceptor adds the cancel function and we're not using this cancel function anyways. I think we could also pass a nil value and have the netceptor do everything.

**Review by fosterseth — APPROVED** [review]

File: pkg/netceptor/netceptor.go
Comment (squidboylan) [inline pkg/netceptor/netceptor.go]: Is this really something we want to export? Im not sure I see the value of that, it seems scary

**Review by squidboylan — APPROVED** [review]
Seems reasonable to me, i left some inline comments

File: pkg/netceptor/netceptor.go
Comment (ghjm) [reply pkg/netceptor/netceptor.go:637]: I think of this as a normal and routine operation that probably doesn't need to spam the logs, but I could be pretty easily talked into this being Info level.  Particularly since it only gets printed when something changes.

File: example/net.go
Comment (ghjm) [reply example/net.go]: Yep, you're right

File: pkg/netceptor/netceptor.go
Comment (ghjm) [reply pkg/netceptor/netceptor.go]: Yep, this is a holdover from the old code and shouldn't exist any more

File: pkg/netceptor/netceptor.go
Comment (squidboylan) [reply pkg/netceptor/netceptor.go:637]: hmm, yeah thinking about it more info is probably too high, but debug seems maybe appropriate. I was mostly just making sure this was intentional. I kinda expected we would want all output to go through the logger unless it cant because the logger isnt setup yet.

Comment (squidboylan) [reply pkg/netceptor/netceptor.go:637]: In any case, we dont need to block on this, we can always change this later without much (any?) pain

Comment (ghjm) [reply pkg/netceptor/netceptor.go:637]: Oh, I see what you're saying.  All the output _is_ going through the logger.  The problem is that the logger adds a newline if one isn't already present, so you can't make multiple logger calls to build a single line the way I was originally doing with the printfs.  Instead, I'm constructing a whole line in the loop before sending it to the logger.  The `fmt.Fprintf` calls are writing _to the `string.Builder`_, not to the console.

**Comment by ghjm** [comment]
@fosterseth I made some more changes to the logger, please have another look

File: pkg/netceptor/netceptor.go
Comment (squidboylan) [reply pkg/netceptor/netceptor.go:637]: Ah! ok, yeah sorry i was confused

**Comment by ghjm** [comment]
> @fosterseth I made some more changes to the logger, please have another look

Actually I have to go so I'm going to merge this, but let me know if you have any objections to the logger changes.

**Pull request merged**
